tests/integration/testlib/iproutelib.py

```diff
@@ -0,0 +1,81 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+import subprocess
+import threading
+import time
+
+
+TIMEOUT = 10
+
+
+class IpMonitorResult(object):
+    def __init__(self):
+        self.out = None
+        self.err = None
+        self.popen = None
+
+
+@contextmanager
+def ip_monitor(object_type, dev, timeout=10):
+    result = IpMonitorResult()
+
+    cmds = 'timeout {} ip monitor {} dev {}'.format(timeout, object_type, dev)
+
+    def run():
+        result.popen = subprocess.Popen(
+            cmds.split(),
+            close_fds=True,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=None
+        )
+        result.out, result.err = result.popen.communicate(None)
+
+    def finalize():
+        if result.popen:
+            result.popen.terminate()
+
+    with _thread(run, 'ip-monitor', teardown_cb=finalize):
+        # Let the ip monitor thread start before proceeding to the action.
+        time.sleep(1)
+        yield result
+
+
+def get_non_up_events(result, dev):
+    """
+    Given a result and device, filter only the non UP events (DOWN, UNKNOWN)
+    and return them as a list.
+    :param result: IpMonitorResult
+    :return: List of non UP events
+    """
+    return [l for l in result.out.split('\n')
```

Comment:
I got error on python 3.
```
TypeError: a bytes-like object is required, not 'str'
```

Reply:
You might use `libcmd.exec_cmd` to save you from handling that.

Reply:
Thanks, fixed.
I could not use `exec_cmd` because I needed the `Popen` object to control the process (which is hidden in `exec_cmd`).